src/java/main/org/apache/zookeeper/server/ZooKeeperServer.java

```diff
@@ -486,19 +475,57 @@ protected void startSessionTracker() {
         ((SessionTrackerImpl)sessionTracker).start();
     }
 
+    /**
+     * Sets the state of ZooKeeper server. After changing the state, this will
+     * notify the server state changes to registered shutdown handler, if any.
+     * <p>
+     * Following are the server state transitions:
+     * <li>During startup the server will be in INITIAL state.</li>
+     * <li>After the successful startup the server will set the state to
+     * RUNNING.</li>
+     * <li>Server will enter into ERROR state if it hits an internal error.
+     * {@link ZooKeeperServerListenerImpl} will notify critical resource error
+     * events.</li>
+     * <li>During shutdown the server will set the state to SHUTDOWN, which
+     * represents that the server is not running.</li>
+     *
+     * @param state new server state.
+     */
+    void setState(State state) {
+        this.state = state;
+        // Notify server state changes to the registered shutdown handler, if any.
+        if (zkShutdownHandler != null) {
```

Comment:
@rakeshadr it sounds good to add the log error for standalone and not for the other cases. You're right that it will add noise to the log if we simply add it here.

I'm not entirely sure why you want to make `setState` protected, though. Don't we need to call it from `notifyStopping`? I'm also not sure why we would make `enum State` protected.